---
title: Exploratory and statistical analysis of RNA-seq data using DESeq2
author: Michael Love
date: June 2017
---

```{r echo=FALSE}
library(knitr)
opts_chunk$set(cache=TRUE)
```

```{r}
library(tximport)
library(readr)
coldata <- read_delim("../data/SraRunTable.txt", delim="\t")
files <- file.path("../data/quant",coldata$Run_s,"quant.sf.gz")
names(files) <- coldata$Run_s
```


```{r eval=FALSE}
# ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_human/release_26/gencode.v26.annotation.gtf.gz
library(GenomicFeatures)
txdb <- makeTxDbFromGFF("gencode.v26.annotation.gtf.gz")
saveDb(txdb, file="gencode.v26.sqlite")
# next time you can just load with this line (no need to makeTxDb...)
# txdb <- loadDb("gencode.v26.sqlite") 
columns(txdb)
k <- keys(txdb, "GENEID")
res <- AnnotationDbi::select(txdb, k, "TXNAME", "GENEID")
tx2gene <- res[,2:1]
```
	
```{r}
load("../data/tx2gene.rda")
```

```{r}
txi <- tximport(files[1], type="salmon", tx2gene=tx2gene)
```

```{r}
txi <- tximport(files, type="salmon", tx2gene=tx2gene)
```

```{r results="hide"}
library(DESeq2)
```

```{r}
geo <- read_delim("../data/GEO_table.txt", delim="\t", col_names=FALSE)
coldata$title <- geo$X2[match(coldata$Sample_Name_s, geo$X1)]
coldata$condition <- factor(coldata$disease_state_s)
coldata$treatment <- factor(coldata$treatment_s)
```

```{r}
dds <- DESeqDataSetFromTximport(txi, coldata,
                                ~condition + treatment + condition:treatment)
```

```{r}
library(magrittr)
levels(dds$condition) <- c("asth","non")
dds$condition %<>% relevel("non")
dds$treatment %<>% relevel("Vehicle")
```

```{r}
vsd <- vst(dds, blind=FALSE)
```

```{r pca}
plotPCA(vsd, c("treatment","condition"))
```

```{r}
dds$id <- substr(dds$title, 1, 3)
id.lvls <- c(dds$id[dds$condition == "non" & dds$treatment == "Vehicle"],
             dds$id[dds$condition == "asth" & dds$treatment == "Vehicle"])
```

```{r}
dds$id %<>% factor(id.lvls)
o <- order(dds$condition, dds$treatment, dds$id)
dds <- dds[,o]
```

```{r}
as.data.frame(colData(dds)[c("condition","treatment","id")])
all(dds$id == c(rep(id.lvls[1:6], 2),
                rep(id.lvls[7:12], 2)))
```

```{r}
dds$id.nested <- factor(rep(1:6,4))
as.data.frame(colData(dds)[c("condition","treatment","id","id.nested")])
```

```{r}
design(dds) <- ~condition + condition:id.nested +
  treatment + condition:treatment
```

```{r}
rownames(dds) <- make.unique(substr(rownames(dds),1,15))
```

```{r}
dds <- dds[rowSums(counts(dds)) > 0,]
keep <- rowSums(counts(dds) >= 5) >= 3
table(keep)
dds <- dds[keep,]
```

```{r}
dds <- DESeq(dds)
resultsNames(dds)
res <- results(dds)
res.sort <- res[order(res$pvalue),]
```

```{r plotma}
plotMA(res, ylim=c(-5,5))
```

```{r}
summary(res)
```

```{r echo=FALSE}
# to make plotCounts same each time
# (has random jitter)
# avoids inflation of git repo...
set.seed(1)
```

```{r topgene1}
top.gene <- rownames(res.sort)[1]
plotCounts(dds, top.gene, c("condition","treatment"), transform=FALSE)
```

```{r}
library(Homo.sapiens)
Homo.sapiens %>% mapIds(top.gene, "SYMBOL", "ENSEMBL")
go.tab <- Homo.sapiens %>% AnnotationDbi::select(top.gene, "GOID", "ENSEMBL") %>% subset(ONTOLOGY == "BP")
# Homo.sapiens %>% AnnotationDbi::select(go.tab$GOID, "TERM", "GOID")
```

```{r}
target <- c("CCL5","CXCL10","CX3CL1","ACKR4","CDHR3")
target.map <- mapIds(Homo.sapiens, target, "ENSEMBL", "SYMBOL")
target.map
match(target.map, rownames(res.sort))
```

```{r targets}
par(mfrow=c(2,1))
plotCounts(dds, target.map[2], c("condition","treatment"))
plotCounts(dds, target.map[2], c("condition","treatment"), transform=FALSE)
```

```{r}
dat <- plotCounts(dds, target.map[2], c("condition","treatment","id.nested"),
                  returnData=TRUE)
```

```{r targets2}
library(ggplot2)
ggplot(dat, aes(x=treatment, y=count, col=id.nested, group=id.nested)) +
  geom_point() + geom_smooth(method="lm") +
  scale_y_log10() + 
  facet_wrap(~condition)
```

```{r plotdisp}
plotDispEsts(dds)
```

```{r}
dispersionFunction(dds)
dmr <- function(x) (.05 + 2.06 / x) * exp(rnorm(length(x),0,.28))
baseMean <- mcols(dds)$baseMean
```

```{r simdisp}
plot(baseMean, dmr(baseMean), log="xy", ylim=c(1e-8,10))
```

```{r}
mean(log2(baseMean))
sd(log2(baseMean))
```

```{r histmean}
hist(2^rnorm(1000,mean=4.80,sd=1.66))
```

```{r}
set.seed(1)
sim <- makeExampleDESeqDataSet(n=10000, m=12,
                               betaSD=0.5,
                               interceptMean=4.80,
                               interceptSD=1.66,
                               dispMeanRel=dmr)
keep <- rowSums(counts(sim) >= 5) >= 3
table(keep)
sim <- sim[keep,]
sim <- DESeq(sim)
sim.res <- results(sim, independentFiltering=FALSE, cooksCutoff=FALSE)
```

```{r compma}
par(mfrow=c(1,2))
plotMA(res, xlim=c(1,1e6), ylim=c(-5,5))
plotMA(sim.res, xlim=c(1,1e6), ylim=c(-5,5))
```

```{r}
sim.dat <- data.frame(sig=sim.res$padj < .1,
                      logmean=cut(mcols(sim)$trueIntercept,c(-1,3.3,6.6,10,20)),
                      abs.lfc=cut(abs(mcols(sim)$trueBeta),c(0,.25,.5,1,3)))
library(dplyr) 
sim.tab <- sim.dat %>% group_by(logmean, abs.lfc) %>% summarize(power=mean(sig))
```

```{r power}
library(ggplot2)
ggplot(sim.tab, aes(x=abs.lfc, y=power, col=logmean, group=logmean)) + geom_line()
```

```{r histmean2}
par(mfrow=c(1,1))
de <- rep(c(FALSE,TRUE),c(8000,2000))
hist(2^rnorm(1000,mean=7,sd=2))
```

```{r}
set.seed(1)
sim2 <- makeExampleDESeqDataSet(n=10000, m=12,
                                betaSD=ifelse(de, 0.5, 0),
                                interceptMean=6,
                                interceptSD=3,
                                dispMeanRel=dmr)
sim2 <- DESeq(sim2)
sim2.res <- results(sim2)
thresh <- c(1,5,10,15,20)/100
fdr <- sapply(thresh, function(t) {
  sig <- which(sim2.res$padj < t)
  mean(!de[sig])
})
```

```{r simfdr}
plot(thresh, fdr, ylim=c(0,.3), type="b", col="blue")
abline(0,1)
```

```{r}
dds2 <- removeResults(dds)
design(dds2) <- ~condition + treatment + condition:id.nested
dds2 <- DESeq(dds2)
resultsNames(dds2)
res2 <- results(dds2, name=c("treatment_HRV16_vs_Vehicle"))
res2 <- results(dds2, contrast=c("treatment","HRV16","Vehicle"))
```

```{r plotma2}
plotMA(res2, ylim=c(-10,10))
```

```{r}
summary(res2)
```

```{r}
res2.sort <- res2[order(res2$log2FoldChange, decreasing=TRUE),]
Homo.sapiens %>% mapIds(rownames(res2.sort)[1:40],
                        "SYMBOL", "ENSEMBL")
```

```{r}
match(target.map, rownames(res2.sort))
go.tab <- Homo.sapiens %>% AnnotationDbi::select(rownames(res2.sort)[1],
                                  "GO", "ENSEMBL") %>% subset(ONTOLOGY == "BP")
```

```{r}
library(GO.db)
go.tab2 <- GO.db %>% AnnotationDbi::select(go.tab$GO, "TERM", "GOID")
substr(go.tab2$TERM, 1, 60)
getTerms <- function(n) {
  go.tab <- Homo.sapiens %>% AnnotationDbi::select(rownames(res2.sort)[n],
                                    "GO", "ENSEMBL") %>% subset(ONTOLOGY == "BP")
  go.tab2 <- Homo.sapiens %>% AnnotationDbi::select(go.tab$GO, "TERM", "GOID")
  substr(go.tab2$TERM, 1, 60)
}
```

```{r}
getTerms(2)
getTerms(3)
getTerms(4)
getTerms(5)
getTerms(6)
getTerms(7)
getTerms(8)
```

```{r}
dds3 <- removeResults(dds)
design(dds3) <- ~condition + treatment
```

```{r}
dds3 <- DESeq(dds3)
resultsNames(dds3)
res3 <- results(dds3, lfcThreshold=2)
res3shr <- lfcShrink(dds3, coef=3, res=res3)
```

```{r plotma3}
par(mfrow=c(1,2))
plotMA(res3, ylim=c(-12,12))
rs <- rowSums(counts(dds)[,dds$treatment == "Vehicle"])
with(res3[rs < 12,], points(baseMean, log2FoldChange, cex=2, col="dodgerblue"))
plotMA(res3shr, ylim=c(-12,12))
with(res3shr[rs < 12,], points(baseMean, log2FoldChange, cex=2, col="dodgerblue"))
```
