---
title: "Exploratory and statistical analysis of RNA-seq data using DESeq2"
author: "Michael Love"
date: "June 2017"
abstract: "This workflow covers basic steps in an exploratory and statistical analysis of RNA-seq data, from import of Salmon quantifications, to performing differential gene expression and looking up annotation information about significant genes. The experimental data consists of control and virus treated samples from 6 asthmatic and 6 non-asthmatic individuals, where a given individual has both a control (Vehicle) and treated (HRV16) sample."
output:
  html_document:
    toc: true
    toc_float: true
---

```{r echo=FALSE}
library(knitr)
opts_chunk$set(cache=TRUE, fig.width=5, fig.height=5)
```

# Importing Salmon quant files

We begin this workflow by importing Salmon quantifications which are 
included in this repository under `data/quant`. Relative to the `scripts` 
directory, we can see we have quantifications for 24 samples:

```{r}
list.files("../data/quant/")
```

The layout of a single sample's quantification directory

```{r}
list.files("../data/quant/SRR1565926")
```

We will read in a table we created from the SRA website, which
gives some of the sample information. We call this table `coldata`,
because it provides data about the *columns* of the count matrix
we will be assembling.

```{r}
library(readr)
coldata <- read_delim("../data/SraRunTable.txt", delim="\t")
coldata
```

We have used the run ID (`SRR...`) to keep track of the reads and quantifications,
so we can build a vector which points to our quantification files using
this column of `coldata`. We use `names` to name this vector with the run IDs as well.

```{r}
files <- file.path("../data/quant",coldata$Run_s,"quant.sf.gz")
names(files) <- coldata$Run_s
head(files,2)
```

The following code (not evaluated here) can be used to generate a table
that connects transcripts to genes for summarizing Salmon transcript
quantifications for differential gene expression. We simply
read in the GTF file from the same database that we used for building
the Salmon index (in this case, Gencode version 26), and then pull
out a table with the transcript name for every gene.

```{r eval=FALSE}
# ftp://ftp.sanger.ac.uk/pub/gencode/Gencode_human/release_26/gencode.v26.annotation.gtf.gz
library(GenomicFeatures)
txdb <- makeTxDbFromGFF("gencode.v26.annotation.gtf.gz")
saveDb(txdb, file="gencode.v26.sqlite")
# next time you can just load with this line (no need to makeTxDb...)
# txdb <- loadDb("gencode.v26.sqlite") 
columns(txdb)
k <- keys(txdb, "GENEID")
res <- AnnotationDbi::select(txdb, k, "TXNAME", "GENEID")
tx2gene <- res[,2:1]
```
	
We have prepared this table in advance, and now load it:
	
```{r}
load("../data/tx2gene.rda")
head(tx2gene)
```

Now we can use the `tximport` function to assemble all the quantifications
from the 24 files, and to summarize the abundances, counts and transcript
lengths to the gene level, for use with DESeq2 and other Bioconductor
packages.

It's a good idea to first test on a single quantification file, which we show here:

```{r}
library(rjson)
library(tximport)
txi <- tximport(files[1], type="salmon", tx2gene=tx2gene)
```

Now we can run `tximport` over all the quanfitication files.
We can see we obtain a list of matrices with common dimension:
58219 (the number of genes) x 24 (the number of samples).

```{r message=FALSE}
txi <- tximport(files, type="salmon", tx2gene=tx2gene)
names(txi)
dim(txi$abundance)
dim(txi$counts)
dim(txi$length)
```

Now we load DESeq2 for further steps in the workflow:

```{r message=FALSE}
library(DESeq2)
```

# Assembling the sample info

```{r}
geo <- read_delim("../data/GEO_table.txt", delim="\t", col_names=FALSE)
coldata$title <- geo$X2[match(coldata$Sample_Name_s, geo$X1)]
coldata$condition <- factor(coldata$disease_state_s)
coldata$treatment <- factor(coldata$treatment_s)
```

```{r}
dds <- DESeqDataSetFromTximport(txi, coldata,
                                ~condition + treatment + condition:treatment)
```

```{r}
library(magrittr)
# you can rename levels, but need to use same order as current levels()
levels(dds$condition) <- c("asth","non")
dds$condition %<>% relevel("non")
dds$treatment %<>% relevel("Vehicle")
```

# Exploratory data analysis

```{r}
vsd <- vst(dds, blind=FALSE)
```

```{r pca, fig.width=7, fig.height=4}
plotPCA(vsd, c("treatment","condition"))
```

# Re-arrange sample info

```{r}
dds$id <- substr(dds$title, 1, 3)
id.lvls <- c(dds$id[dds$condition == "non" & dds$treatment == "Vehicle"],
             dds$id[dds$condition == "asth" & dds$treatment == "Vehicle"])
```

```{r}
dds$id %<>% factor(id.lvls)
o <- order(dds$condition, dds$treatment, dds$id)
dds <- dds[,o]
```

```{r}
as.data.frame(colData(dds)[c("condition","treatment","id")])
all(dds$id == c(rep(id.lvls[1:6], 2),
                rep(id.lvls[7:12], 2)))
```

```{r}
dds$id.nested <- factor(rep(1:6,4))
as.data.frame(colData(dds)[c("condition","treatment","id","id.nested")])
```

```{r}
design(dds) <- ~condition + condition:id.nested +
  treatment + condition:treatment
```

```{r}
rownames(dds) <- make.unique(substr(rownames(dds),1,15))
```

# Differential gene expression

```{r}
dds <- dds[rowSums(counts(dds)) > 0,]
keep <- rowSums(counts(dds) >= 10) >= 3
table(keep)
dds <- dds[keep,]
```

```{r}
dds <- DESeq(dds)
resultsNames(dds)
res <- results(dds)
res.sort <- res[order(res$pvalue),]
```

```{r plotma}
plotMA(res, ylim=c(-5,5))
```

```{r}
summary(res)
```

```{r echo=FALSE}
# to make plotCounts same each time
# (has random jitter)
# avoids inflation of git repo...
set.seed(1)
```

```{r topgene1}
top.gene <- rownames(res.sort)[1]
plotCounts(dds, top.gene, c("condition","treatment"), transform=FALSE)
```

# Exploring results

```{r}
library(Homo.sapiens)
Homo.sapiens %>% mapIds(top.gene, "SYMBOL", "ENSEMBL")
go.tab <- Homo.sapiens %>% AnnotationDbi::select(top.gene, "GOID", "ENSEMBL") %>% subset(ONTOLOGY == "BP")
# Homo.sapiens %>% AnnotationDbi::select(go.tab$GOID, "TERM", "GOID")
```

```{r}
target <- c("CCL5","CXCL10","CX3CL1","ACKR4","CDHR3")
target.map <- mapIds(Homo.sapiens, target, "ENSEMBL", "SYMBOL")
target.map
match(target.map, rownames(res.sort))
```

```{r targets}
plotCounts(dds, target.map[2], c("condition","treatment"))
plotCounts(dds, target.map[2], c("condition","treatment"), transform=FALSE)
```

```{r}
dat <- plotCounts(dds, target.map[2], c("condition","treatment","id.nested"),
                  returnData=TRUE)
```

```{r targets2, warning=FALSE, fig.width=7, fig.height=4}
library(ggplot2)
ggplot(dat, aes(x=treatment, y=count, col=id.nested, group=id.nested)) +
  geom_point() + geom_smooth(method="lm") +
  scale_y_log10() + 
  facet_wrap(~condition)
```

# Interlude: power analysis

```{r plotdisp}
plotDispEsts(dds, ylim=c(1e-4,10))
```

```{r}
dispersionFunction(dds)
dmr <- function(x) (.05 + 2.06 / x) * exp(rnorm(length(x),0,.28))
baseMean <- mcols(dds)$baseMean
```

```{r simdisp}
plot(baseMean, dmr(baseMean), log="xy", ylim=c(1e-4,10))
```

```{r}
mean(log2(baseMean))
sd(log2(baseMean))
```

```{r}
set.seed(1)
sim <- makeExampleDESeqDataSet(n=10000, m=12,
                               betaSD=1,
                               interceptMean=5,
                               interceptSD=2,
                               dispMeanRel=dmr)
keep <- rowSums(counts(sim) >= 10) >= 3
table(keep)
sim <- sim[keep,]
sim <- DESeq(sim)
sim.res <- results(sim, independentFiltering=FALSE, cooksCutoff=FALSE)
```

```{r compma, fig.width=8, fig.height=5}
par(mfrow=c(1,2))
plotMA(res, xlim=c(1,1e6), ylim=c(-5,5))
plotMA(sim.res, xlim=c(1,1e6), ylim=c(-5,5))
```

```{r}
max.lfc <- ceiling(max(abs(mcols(sim)$trueBeta)))
sim.dat <- data.frame(sig=sim.res$padj < .1,
                      log2mean=cut(mcols(sim)$trueIntercept,c(-1,3.3,6.6,10,20)),
                      abs.lfc=cut(abs(mcols(sim)$trueBeta),c(0,.25,.5,1,max.lfc)))
```

```{r message=FALSE}
library(dplyr) 
sim.tab <- sim.dat %>% group_by(log2mean, abs.lfc) %>% summarize(power=mean(sig))
```

```{r power}
ggplot(sim.tab, aes(x=abs.lfc, y=power, col=log2mean, group=log2mean)) + geom_line()
```

# Empirical and nominal FDR

```{r}
de <- rep(c(FALSE,TRUE),c(8000,2000))
set.seed(1)
sim2 <- makeExampleDESeqDataSet(n=10000, m=12,
                                betaSD=ifelse(de,1,0),
                                interceptMean=6,
                                interceptSD=3,
                                dispMeanRel=dmr)
sim2 <- DESeq(sim2)
threshold <- c(1,5,10,15,20)/100
FDR <- sapply(threshold, function(t) {
  sim2.res <- results(sim2, alpha=t)
  sig <- which(sim2.res$padj < t)
  mean(!de[sig])
})
```

```{r simfdr}
plot(threshold, FDR, ylim=c(0,.3), type="b", col="blue",
     main="Empirical vs nominal FDR")
abline(0,1)
```

# Other differential analyses

```{r}
dds2 <- removeResults(dds)
design(dds2) <- ~condition + treatment + condition:id.nested
dds2 <- DESeq(dds2)
resultsNames(dds2)
res2 <- results(dds2, name=c("treatment_HRV16_vs_Vehicle"))
res2 <- results(dds2, contrast=c("treatment","HRV16","Vehicle"))
```

```{r plotma2}
plotMA(res2, ylim=c(-10,10))
```

```{r}
summary(res2)
```

```{r}
res2.sort <- res2[order(res2$log2FoldChange, decreasing=TRUE),]
Homo.sapiens %>% mapIds(rownames(res2.sort)[1:40],
                        "SYMBOL", "ENSEMBL")
```

# Exploring results with annotation

```{r}
match(target.map, rownames(res2.sort))
go.tab <- Homo.sapiens %>% AnnotationDbi::select(rownames(res2.sort)[1],
                                  "GO", "ENSEMBL") %>% subset(ONTOLOGY == "BP")
```

```{r}
library(GO.db)
go.tab2 <- GO.db %>% AnnotationDbi::select(go.tab$GO, "TERM", "GOID")
substr(go.tab2$TERM, 1, 60)
getTerms <- function(n) {
  go.tab <- Homo.sapiens %>% AnnotationDbi::select(rownames(res2.sort)[n],
                                    "GO", "ENSEMBL") %>% subset(ONTOLOGY == "BP")
  go.tab2 <- Homo.sapiens %>% AnnotationDbi::select(go.tab$GO, "TERM", "GOID")
  substr(go.tab2$TERM, 1, 60)
}
```

```{r}
getTerms(2)
getTerms(3)
getTerms(4)
getTerms(5)
getTerms(6)
getTerms(7)
getTerms(8)
```

# LFC shrinkage

```{r}
dds3 <- removeResults(dds)
design(dds3) <- ~condition + treatment
```

```{r}
dds3 <- DESeq(dds3)
resultsNames(dds3)
res3 <- results(dds3, lfcThreshold=2)
res3shr <- lfcShrink(dds3, coef=3, res=res3)
```

```{r plotma3, fig.width=8, fig.height=5}
par(mfrow=c(1,2))
plotMA(res3, ylim=c(-12,12))
rs <- rowSums(counts(dds)[,dds$treatment == "Vehicle"])
with(res3[rs < 12,], points(baseMean, log2FoldChange, cex=2, col="dodgerblue"))
plotMA(res3shr, ylim=c(-12,12))
with(res3shr[rs < 12,], points(baseMean, log2FoldChange, cex=2, col="dodgerblue"))
```

```{r}
sessionInfo()
```
